File: python/src/bursar/credits/postgres/repositories/plan.py
from __future__ import annotations

from uuid import UUID

from bursar.credits.postgres.repositories._types import DbQuery
from bursar.credits.postgres.repositories._utils import (
    optional_mapping_row,
    require_mapping_row,
    require_row,
    validate_non_empty,
    validate_row,
)
from bursar.credits.postgres.repositories.schemas import (
    AllowanceRow,
    EntitlementRow,
    PlanMigrationBatchRow,
    QuotaEventRow,
    QuotaStateRow,
    SetUserPlanRow,
    UnsetUserPlanRow,
    UserPlanRow,
)
from bursar.errors import StoreError


class PlanRepository:
    """Repository for user plan operations.

    All methods call Postgres RPCs via the callproc function.
    Returns None when the RPC returns no rows (no result).
    Returns typed Pydantic models for successful results.
    """

    def __init__(self, callproc: DbQuery, query: DbQuery | None = None) -> None:
        self._callproc = callproc
        del query

    def get_user_plan(self, user_id: str) -> UserPlanRow | None:
        """Get the current plan for a user.

        Args:
            user_id: The user ID.

        Returns:
            UserPlanRow if found, None if the user has no plan assigned.
        """
        validate_non_empty(user_id, "user_id")
        row = optional_mapping_row(
            self._callproc("get_subject_plan", [user_id]),
            "PlanRepository.get_user_plan",
        )
        if row is None:
            return None
        return validate_row(UserPlanRow, row, "PlanRepository.get_user_plan")

    def get_entitlement(self, user_id: str, feature: str) -> EntitlementRow | None:
        validate_non_empty(user_id, "user_id")
        validate_non_empty(feature, "feature")
        rows = self._callproc("get_subject_entitlements", [user_id]) or []
        entitlements = [validate_row(EntitlementRow, row, "PlanRepository.get_entitlement") for row in rows]
        return next((row for row in entitlements if row.feature_key == feature), None)

    def set_user_plan(
        self,
        user_id: str,
        plan_key: str,
        plan_assigned_at: str | None,
    ) -> SetUserPlanRow:
        """Assign a plan to a user.

        Args:
            user_id: The user ID.
            plan_key: The public plan key.
            plan_assigned_at: ISO datetime string for the assignment, or None.

        Returns:
            SetUserPlanRow if successful, None if the RPC returned no rows.
        """
        validate_non_empty(user_id, "user_id")
        validate_non_empty(plan_key, "plan_key")
        row = require_mapping_row(
            self._callproc("set_subject_plan", [user_id, plan_key, plan_assigned_at]),
            "PlanRepository.set_user_plan",
        )
        return validate_row(
            SetUserPlanRow,
            row,
            "PlanRepository.set_user_plan",
            indeterminate=True,
        )

    def unset_user_plan(self, user_id: str) -> UnsetUserPlanRow:
        """Remove the plan assignment from a user.

        Args:
            user_id: The user ID.

        Returns:
            UnsetUserPlanRow if successful, None if the user had no plan.
        """
        validate_non_empty(user_id, "user_id")
        rows = self._callproc("unassign_plan", [user_id, "sdk_unassignment"])
        result = require_row(rows, "PlanRepository.unset_user_plan")
        if type(result) is not bool:
            raise StoreError(
                "PlanRepository.unset_user_plan: expected a boolean result",
                indeterminate=True,
            )
        if not result:
            raise StoreError("PlanRepository.unset_user_plan: unassign_plan returned false")
        return validate_row(
            UnsetUserPlanRow,
            {"user_id": user_id},
            "PlanRepository.unset_user_plan",
        )

    def set_plan_revision_pin(self, user_id: str, pinned: bool) -> bool:
        """Pin or unpin the user's current assignment to its catalog revision."""
        validate_non_empty(user_id, "user_id")
        rows = self._callproc("set_plan_revision_pin", [user_id, pinned])
        result = require_row(rows, "PlanRepository.set_plan_revision_pin")
        if type(result) is not bool:
            raise StoreError("PlanRepository.set_plan_revision_pin: expected a boolean result")
        return result

    def apply_due_plan_changes(self, limit: int) -> int:
        """Apply one bounded batch of renewal-effective plan changes."""
        rows = self._callproc("apply_due_plan_assignment_changes", [limit]) or []
        result = require_row(rows, "PlanRepository.apply_due_plan_changes")
        if isinstance(result, bool) or not isinstance(result, int) or result < 0:
            raise StoreError("PlanRepository.apply_due_plan_changes: expected a non-negative integer")
        return result

    def start_plan_migration(
        self,
        from_plan_id: str | None,
        to_plan_id: str,
    ) -> str:
        """Start a resumable migration between catalog plan records.

        Args:
            from_plan_id: Source catalog plan ID, or None for all assignments.
            to_plan_id: Target catalog plan ID.

        Returns:
            Migration ID.
        """
        if from_plan_id is not None:
            validate_non_empty(from_plan_id, "from_plan_id")
        validate_non_empty(to_plan_id, "to_plan_id")
        rows = self._callproc("start_plan_migration", [from_plan_id, to_plan_id])
        value = require_row(rows, "PlanRepository.start_plan_migration")
        try:
            return str(UUID(str(value)))
        except (AttributeError, TypeError, ValueError) as error:
            raise StoreError(
                "PlanRepository.start_plan_migration: expected a UUID result",
                indeterminate=True,
            ) from error

    def migrate_plan_batch(
        self,
        migration_id: str,
        batch_size: int = 100,
    ) -> PlanMigrationBatchRow:
        """Advance a plan migration by one bounded batch."""
        validate_non_empty(migration_id, "migration_id")
        if batch_size < 1 or batch_size > 1000:
            raise ValueError("batch_size must be between 1 and 1000")
        row = require_mapping_row(
            self._callproc("migrate_plan_batch", [migration_id, batch_size]),
            "PlanRepository.migrate_plan_batch",
        )
        return validate_row(
            PlanMigrationBatchRow,
            row,
            "PlanRepository.migrate_plan_batch",
            indeterminate=True,
        )

    def get_quota_state(
        self,
        user_id: str,
        quota_key: str | None = None,
    ) -> list[QuotaStateRow]:
        validate_non_empty(user_id, "user_id")
        rows = self._callproc(
            "get_subject_quota_state",
            [user_id, quota_key],
        )
        return [validate_row(QuotaStateRow, row, "PlanRepository.get_quota_state") for row in rows]

    def check_allowance(self, user_id: str) -> AllowanceRow | None:
        """Check the remaining plan allowance for a user.

        Args:
            user_id: The user ID.
        Returns:
            AllowanceRow if found, None if no plan or allowance configured.
        """
        validate_non_empty(user_id, "user_id")
        row = optional_mapping_row(
            self._callproc("get_subject_allowance", [user_id]),
            "PlanRepository.check_allowance",
        )
        if row is None:
            return None
        return validate_row(AllowanceRow, row, "PlanRepository.check_allowance")

    def list_quota_events(
        self,
        user_id: str,
        after: str | None,
        limit: int,
        idempotency_key: str | None,
        after_id: str | None,
    ) -> list[QuotaEventRow]:
        validate_non_empty(user_id, "user_id")
        rows = self._callproc(
            "list_subject_quota_events",
            [user_id, after, limit, idempotency_key, after_id],
        )
        return [validate_row(QuotaEventRow, row, "PlanRepository.list_quota_events") for row in rows]

File: javascript/src/credits/postgres/repositories/plan.ts
import { z } from "zod";
import type { CallProc } from "../../../shared/postgres-types.js";
import {
  optionalRecordRow,
  postgresUuid,
  requireRow,
  safeParse,
} from "../../../shared/postgres-validation.js";
import { StoreError } from "../../../errors.js";

const decimal = z.union([z.string().min(1), z.number().finite()] as const);
const timestamp = z.union([z.string().min(1), z.date()] as const).transform((value, context) => {
  const parsed = value instanceof Date ? value : new Date(value);
  if (Number.isNaN(parsed.getTime())) {
    context.addIssue({ code: "custom", message: "expected a valid timestamp" });
    return z.NEVER;
  }
  return parsed;
});
const safeInteger = z
  .union([z.number().int(), z.string().regex(/^\d+$/)] as const)
  .transform(Number)
  .refine(Number.isSafeInteger, "expected a safe integer");
const entitlementValue = z.object({ value: z.json() }).strict();
const admissionOperation = z
  .object({ max_in_flight: z.number().int().positive().nullable() })
  .strict();

const UserPlanRowSchema = z
  .object({
    user_id: postgresUuid,
    plan_assigned_at: timestamp,
    plan_assignment_ends_at: timestamp.nullable(),
    assignment_source_type: z.enum(["manual", "subscription", "migration", "system"]),
    assignment_source_id: postgresUuid.nullable(),
    catalog_revision_pinned: z.boolean(),
    plan_id: postgresUuid,
    plan_key: z.string().min(1),
    plan_label: z.string().min(1),
    rate_card: z.string().min(1).nullable(),
    allowed_operations: z.array(z.string().min(1)),
    credit_allowance_amount: decimal.nullable(),
    credit_allowance_priority: z.number().int().nonnegative().nullable(),
    credit_allowance_reset_unit: z
      .enum(["second", "minute", "hour", "day", "week", "month", "year"])
      .nullable(),
    credit_allowance_reset_count: z.number().int().positive().nullable(),
    credit_allowance_reset_anchor: z.enum(["calendar", "plan_assignment", "rolling"]).nullable(),
    credit_allowance_reset_timezone: z.string().min(1).nullable(),
    entitlements: z.record(z.string().min(1), entitlementValue),
    credit_policy_type: z.enum(["prepaid", "credit_line"]).nullable(),
    credit_limit: decimal.nullable(),
    admission_max_in_flight: z.number().int().positive().nullable(),
    operation_admission: z.record(z.string().min(1), admissionOperation),
    catalog_revision_no: safeInteger,
  })
  .strict()
  .superRefine((row, context) => {
    const allowance = [
      row.credit_allowance_amount,
      row.credit_allowance_priority,
      row.credit_allowance_reset_unit,
      row.credit_allowance_reset_count,
      row.credit_allowance_reset_anchor,
      row.credit_allowance_reset_timezone,
    ];
    const populated = allowance.filter((value) => value !== null).length;
    if (populated !== 0 && populated !== allowance.length) {
      context.addIssue({
        code: "custom",
        message: "allowance policy fields must be all set or all null",
      });
    }
    if (
      (row.credit_policy_type === null && row.credit_limit !== null) ||
      (row.credit_policy_type === "prepaid" && row.credit_limit !== null) ||
      (row.credit_policy_type === "credit_line" && row.credit_limit === null)
    ) {
      context.addIssue({ code: "custom", message: "credit policy fields are inconsistent" });
    }
  });

const SetUserPlanRowSchema = z
  .object({
    user_id: postgresUuid,
    plan_id: postgresUuid,
    plan_key: z.string().min(1),
    plan_assigned_at: timestamp,
    assignment_state: z.enum(["applied", "scheduled"]),
  })
  .strict();

const EntitlementRowSchema = z
  .object({
    feature_key: z.string().min(1),
    feature_type: z.enum(["boolean", "integer", "string", "enum"]),
    feature_value: z.json(),
    catalog_revision_id: postgresUuid,
    plan_key: z.string().min(1).nullable(),
    value_source: z.enum(["default", "plan"]),
  })
  .strict();

const AllowanceRowSchema = z
  .object({
    plan_id: postgresUuid,
    allowance_remaining: decimal,
    period_start: timestamp.transform((value) => value.toISOString()),
    period_end: timestamp.transform((value) => value.toISOString()),
  })
  .strict();

const UnsetPlanRowSchema = z
  .object({
    user_id: postgresUuid,
  })
  .strict();

const PlanMigrationStartRowSchema = postgresUuid;
const PlanMigrationBatchRowSchema = z
  .object({
    migrated: z.number().int().nonnegative(),
    done: z.boolean(),
    next_cursor: postgresUuid.nullable(),
  })
  .strict();
const QuotaStateRowSchema = z
  .object({
    user_id: postgresUuid,
    quota_key: z.string().min(1),
    operation_key: z.string().min(1),
    measure_key: z.string().min(1),
    quota_limit: decimal,
    consumed: decimal,
    reserved: decimal,
    remaining: decimal,
    overage: decimal,
    enforcement: z.enum(["block", "allow"]),
    window_start: timestamp.transform((value) => value.toISOString()),
    window_end: timestamp.transform((value) => value.toISOString()),
    emit_at_percent: z.array(z.number().finite().min(0).max(100)),
  })
  .strict();
const QuotaEventRowSchema = z
  .object({
    event_id: postgresUuid,
    quota_key: z.string().min(1),
    operation_key: z.string().min(1),
    measure_key: z.string().min(1),
    event_type: z.enum(["threshold", "blocked"]),
    threshold_percent: z.number().finite().min(0).max(100).nullable(),
    idempotency_key: z.string().min(1),
    usage_charge_id: postgresUuid.nullable(),
    created_at: timestamp.transform((value) => value.toISOString()),
  })
  .strict();

export type UnsetPlanRow = z.infer<typeof UnsetPlanRowSchema>;
export type UserPlanRow = z.infer<typeof UserPlanRowSchema>;
export type SetUserPlanRow = z.infer<typeof SetUserPlanRowSchema>;
export type EntitlementRow = z.infer<typeof EntitlementRowSchema>;
export type AllowanceRow = z.infer<typeof AllowanceRowSchema>;
export type PlanMigrationStartRow = z.infer<typeof PlanMigrationStartRowSchema>;
export type PlanMigrationBatchRow = z.infer<typeof PlanMigrationBatchRowSchema>;
export type QuotaStateRow = z.infer<typeof QuotaStateRowSchema>;
export type QuotaEventRow = z.infer<typeof QuotaEventRowSchema>;

/** Repository for plan management operations. */
export class PlanRepository {
  constructor(private callproc: CallProc) {}

  /** Fetch a user's current plan. Returns null if user has no plan. */
  async getUserPlan(userId: string): Promise<UserPlanRow | null> {
    const rows = await this.callproc("get_subject_plan", [userId]);
    const row = optionalRecordRow(rows, "PlanRepository.getUserPlan");
    return row === null ? null : safeParse(UserPlanRowSchema, row, "PlanRepository.getUserPlan");
  }

  async getEntitlement(userId: string, feature: string): Promise<EntitlementRow | null> {
    const rows = await this.callproc("get_subject_entitlements", [userId]);
    const entitlements = (rows ?? []).map((row) =>
      safeParse(EntitlementRowSchema, row, "PlanRepository.getEntitlement"),
    );
    return entitlements.find((candidate) => candidate.feature_key === feature) ?? null;
  }

  /** Assign a plan to a user. */
  async setUserPlan(
    userId: string,
    planKey: string,
    planAssignedAt: string | null,
  ): Promise<SetUserPlanRow> {
    return safeParse(
      SetUserPlanRowSchema,
      requireRow(
        await this.callproc("set_subject_plan", [userId, planKey, planAssignedAt]),
        "PlanRepository.setUserPlan",
      ),
      "PlanRepository.setUserPlan",
    );
  }

  /** Remove a user's plan assignment. */
  async unsetUserPlan(userId: string): Promise<UnsetPlanRow> {
    const rows = await this.callproc("unassign_plan", [userId, "sdk_unassignment"]);
    if (
      !safeParse(
        z.boolean(),
        requireRow(rows, "PlanRepository.unsetUserPlan"),
        "PlanRepository.unsetUserPlan",
      )
    ) {
      throw new StoreError("PlanRepository.unsetUserPlan: unassign_plan returned false");
    }
    return safeParse(UnsetPlanRowSchema, { user_id: userId }, "PlanRepository.unsetUserPlan");
  }

  /** Pin or unpin the user's current assignment to its catalog revision. */
  async setPlanRevisionPin(userId: string, pinned: boolean): Promise<boolean> {
    const rows = await this.callproc("set_plan_revision_pin", [userId, pinned]);
    return safeParse(
      z.boolean(),
      requireRow(rows, "PlanRepository.setPlanRevisionPin"),
      "PlanRepository.setPlanRevisionPin",
    );
  }

  /** Apply one bounded batch of renewal-effective plan changes. */
  async applyDuePlanChanges(limit: number): Promise<number> {
    const rows = await this.callproc("apply_due_plan_assignment_changes", [limit]);
    return safeParse(
      z.number().int().nonnegative(),
      requireRow(rows, "PlanRepository.applyDuePlanChanges"),
      "PlanRepository.applyDuePlanChanges",
    );
  }

  /** Check a user's remaining free allowance. */
  async startPlanMigration(
    fromPlanId: string | null,
    toPlanId: string,
  ): Promise<PlanMigrationStartRow> {
    const rows = await this.callproc("start_plan_migration", [fromPlanId, toPlanId]);
    return safeParse(
      PlanMigrationStartRowSchema,
      requireRow(rows, "PlanRepository.startPlanMigration"),
      "PlanRepository.startPlanMigration",
    );
  }

  async migratePlanBatch(migrationId: string, batchSize?: number): Promise<PlanMigrationBatchRow> {
    const rows = await this.callproc("migrate_plan_batch", [migrationId, batchSize ?? 100]);
    return safeParse(
      PlanMigrationBatchRowSchema,
      requireRow(rows, "PlanRepository.migratePlanBatch"),
      "PlanRepository.migratePlanBatch",
    );
  }

  async getQuotaState(userId: string, quotaKey?: string | null): Promise<QuotaStateRow[]> {
    const rows = await this.callproc("get_subject_quota_state", [userId, quotaKey ?? null]);
    return (rows ?? []).map((row) =>
      safeParse(QuotaStateRowSchema, row, "PlanRepository.getQuotaState"),
    );
  }

  async listQuotaEvents(
    userId: string,
    after: string | null,
    limit: number,
    idempotencyKey: string | null,
    afterId: string | null,
  ): Promise<QuotaEventRow[]> {
    const rows = await this.callproc("list_subject_quota_events", [
      userId,
      after,
      limit,
      idempotencyKey,
      afterId,
    ]);
    return (rows ?? []).map((row) =>
      safeParse(QuotaEventRowSchema, row, "PlanRepository.listQuotaEvents"),
    );
  }

  async checkAllowance(userId: string): Promise<AllowanceRow | null> {
    const rows = await this.callproc("get_subject_allowance", [userId]);
    const row = optionalRecordRow(rows, "PlanRepository.checkAllowance");
    return row === null
      ? null
      : safeParse(AllowanceRowSchema, row, "PlanRepository.checkAllowance");
  }
}
